File: src/main/java/com/vc/medium/L61.scala
package com.vc.medium

class ListNode61(var _x: Int = 0) {
  var next: ListNode61 = null
  var x: Int = _x
}

object L61 {
  def rotateRight(head: ListNode61, k: Int): ListNode61 = {
    if(head == null || head.next == null) return head

    var n = 0
    val dummyNode = new ListNode61()
    dummyNode.next = head
    var hare = dummyNode
    var tortoise = dummyNode
    while(hare.next != null) {
      n += 1
      hare = hare.next
    }

    var i = 0
    while(i < n - k % n) {
      tortoise = tortoise.next
      i += 1
    }

    hare.next = dummyNode.next
    dummyNode.next = tortoise.next
    tortoise.next = null
    dummyNode.next
  }
}

File: src/main/java/com/vc/medium/L61.java
package com.vc.medium;

import com.vc.hard.ListNode;

class L61 {
    public ListNode rotateRight(ListNode head, int k) {
        if(head == null || head.next == null) return head;

        ListNode tail = head;
        int size = 0;
        while(tail.next != null) {
            tail = tail.next;
            size++;
        }

        //Make it circular
        tail.next = head;

        size++;
        //Run loop k times to get to next head
        tail = head;
        int count = size - k % size - 1;
        while(count > 0) {
            tail = tail.next;
            count--;
        }

        ListNode toReturn = tail.next;
        tail.next = null;

        return toReturn;
    }
}
